File: 2023/1/main.ml

let reverse str = 
  let rec loop idx = match idx with
    0 -> String.make 1 str.[0]
    | _ -> String.make 1 str.[idx] ^ loop (idx - 1)
  in loop (String.length str - 1)
;;

let digit_word str idx =
  let words = ["one"; "two"; "three"; "four"; "five"; "six"; "seven"; "eight"; "nine"] in
  let str = String.sub str idx (String.length str - idx) in
    let _ = Printf.printf "%s\n" str in
    let opt_word i word = 
      if String.starts_with ~prefix:word str then
        Some (i + 1)
      else
        None
    in
  
    let matches = List.filter_map Fun.id (List.mapi opt_word words) in
      match matches with
        [] -> None
        | item :: rest -> Some item
;;

let first_digit_or_word str =
  let rec loop idx =
    if idx < String.length str then
      let c = str.[idx] in
        match c with
          '0' .. '9' -> Some (int_of_string (String.make 1 c))
          | _ -> match digit_word str idx with
            Some s -> Some s
            | None -> loop (idx + 1)
    else
      None

  in loop 0
;;

let last_digit_or_word str =
  let rec loop idx =
    if idx >= 0 then
      let c = str.[idx] in
        match c with
          '0' .. '9' -> Some (int_of_string (String.make 1 c))
          | _ -> match digit_word str idx with
            Some s -> Some s
            | None -> loop (idx - 1)
    else
      None

  in loop ((String.length str) - 1)
;;
  
  
let first_digit str =
  let rec loop idx =
    if idx < String.length str then
      let c = str.[idx] in
        match c with
          '0' .. '9' -> Some (c, idx)
          | _ -> loop (idx + 1)
    else
      None

  in loop 0
;;


let first_and_last_digit str =
  let first = first_digit_or_word str in
    let second = last_digit_or_word str in
      match first, second with
        Some first, Some second -> Some (first * 10 + second)
        | _ -> None
;;

let map_file fn path =
  let f = open_in path in
    let rec loop () =
      try
        let next = input_line f in
        fn(next) :: loop ()
      with End_of_file ->
        close_in f;
        []
    in loop ()
;;

let option_to_int = function
  | None -> 0
  | Some s -> s
;;

let remove_nones l = List.map option_to_int l;;

let rec sum_list = function
  [] -> 0
  | head :: tail -> head + (sum_list tail)
;;

let arr = map_file first_and_last_digit "input.txt" in
  let sum = sum_list (remove_nones arr) in
    print_int sum
;;

File: 2023/1/rust/src/main.rs
use std::fs;

const WORD_MAP: [(&str, u32); 18] = [
    ("one", 1),
    ("1", 1),
    ("two", 2),
    ("2", 2),
    ("three", 3),
    ("3", 3),
    ("four", 4),
    ("4", 4),
    ("five", 5),
    ("5", 5),
    ("six", 6),
    ("6", 6),
    ("seven", 7),
    ("7", 7),
    ("eight", 8),
    ("8", 8),
    ("nine", 9),
    ("9", 9),
];

fn first_digit_or_word(line: &[u8], mut indices: impl Iterator<Item = usize>) -> Option<u32> {
    indices.find_map(|s| {
        WORD_MAP
            .iter()
            .find_map(|(word, value)| line[s..].starts_with(word.as_bytes()).then_some(*value))
    })
}

fn find_value(line: &[u8]) -> u32 {
    let range = 0..line.len();
    let first = first_digit_or_word(line, range.clone());
    let last = first_digit_or_word(line, range.rev());

    first.unwrap() * 10 + last.unwrap()
}

fn main() {
    let result: u32 = fs::read_to_string("../input.txt")
        .unwrap()
        .lines()
        .map(|l| find_value(l.as_bytes()))
        .sum();

    println!("{result}");
}
